### ISSUE TITLE
Any Progress bar?

### ISSUE THREAD

**Issue by ohld** [nkaz001/hftbacktest#38]
TL;DR:
Feature request: progress bar for backtesting.

What I did:
1. Downloaded data using https://github.com/nkaz001/collect-binancefutures
2. Tried to run https://github.com/nkaz001/hftbacktest/blob/master/examples/Market%20Making%20with%20Alpha%20-%20Order%20Book%20Imbalance.ipynb on my dataset.
3. `obi_mm(hbt, ...)` became silent after the data was loaded.

So I don't understand if the backtest is even working. And what if it would require several days to backtest on my laptop?

It would be great to include a progress bar somewhere. `tqdm` package is a great tool to have pretty progressbars in raw python and in jupyter environments.

**Comment by nkaz001** [comment]
I will consider showing progress. But, due to Numba limitations, displaying a progress bar might be challenging. Printing progress is acceptable, even if it might be a bit verbose. In the meantime, you can display progress as follows:

    progress = 0
    print('Start')
    while hbt.elapse(interval):
        if hbt.local.data is not None:
            progress_update = int(hbt.local.row_num / len(hbt.local.data) * 100)
            if progress_update != progress:
                print(progress_update, '%')
                progress = progress_update

**Comment by alihaskar** [comment]
Hi @ohld and @nkaz001 ,

Numba kernels can’t update tqdm directly, but you can still see rough progress by printing a snapshot every few simulated hours.
Here’s a minimal template that works inside njit (nopython=True) and keeps the hot loop fast:

```

@njit
def run_with_progress(hbt):
    STEP_NS        = 100_000_000                 # 100 ms per engine step
    PRINT_EVERY_NS = 4 * 60 * 60 * 1_000_000_000 # 4 h of *simulated* time
    asset_no        = 0
    start_time      = hbt.current_timestamp
    last_print_time = start_time

    # main replay loop --------------------------------------------------
    while hbt.elapse(STEP_NS) == 0:          # stop when elapse() returns 1
        hbt.clear_inactive_orders(asset_no)

        # ---- strategy logic here ----
        # your_strategy_step(hbt)
        # -----------------------------

        now = hbt.current_timestamp
        if now - last_print_time >= PRINT_EVERY_NS:
            hours  = (now - start_time) / (60.0 * 60.0 * 1_000_000_000)
            days   = hours / 24.0
            state  = hbt.state_values(asset_no)

            # NOTE: Numba doesn’t support f-strings – use %-formatting
            print("Progress: %.1f days elapsed" % days)
            print("Timestamp (ns): %d"       % now)
            print("Balance: %.2f"            % state.balance)
            print("Fee: %.2f"                % state.fee)
            print("Position: %.4f"           % hbt.position(asset_no))
            print("----------------------------------------")

            last_print_time = now
```
